animal_crossing.py

```diff
@@ -0,0 +1,70 @@
+import re
+import sqlite3
+
+import config
+
+if config.bot.acnh_db is not None:
+	db = sqlite3.connect(config.bot.acnh_db)
+	db.row_factory = sqlite3.Row
+
+def friend_code(cmd):
+	if not cmd.args:
+		_user_list_all(cmd)
+		return
+
+	split = cmd.args.split(' ', 1)
+	subcmd = split[0]
+	if subcmd == 'set':
+		_user_upsert_friend_code(cmd, split[1])
+	elif subcmd == 'remove':
+		_user_remove(cmd)
+	else:
+		_user_find(cmd, split[0])
+
+def _user_upsert_friend_code(cmd, code):
+	pat = re.compile(r'SW-\d{4}-\d{4}-\d{4}')
+	if pat.match(code) is None:
+		cmd.reply('Invalid friend code submitted.')
+		return
+
+	sender = cmd.sender
+	db.execute('''
+	INSERT INTO user VALUES(?, ?, ?)
+	ON CONFLICT(id)
+	DO UPDATE SET code=excluded.code
+	''', (sender['id'], sender['username'], code))
+	db.commit()
+	cmd.reply('Friend code for %s has been set.' % (sender['username']))
+
+def _user_list_all(cmd):
+	cur = db.execute('SELECT username, code FROM user')
+	try:
+		users = cur.fetchall()
+		reply = []
+		for user in users:
+			reply.append('%s: %s' % (user['username'], user['code']))
+		cmd.reply('\n'.join(reply))
+	except sqlite3.OperationalError:
+		cmd.reply('There are no friend codes saved :(')
+
+def _user_find(cmd, user):
+	cur = db.execute('''
+	SELECT username, code FROM user WHERE username LIKE ?
+	''', (user,))
+	try:
+		user = cur.fetchone()
+		cmd.reply('%s: %s' % (user[0], user[1]))
+	except sqlite3.OperationalError:
+		cmd.reply('Friend code for %s could not be found.' % (user))
+
+def _user_remove(cmd):
+	db.execute('''
+	DELETE FROM user WHERE id=?
+	''', (cmd.sender['id'],))
+
+	try:
+		db.commit()
```

Comment:
you can https://docs.python.org/3/library/sqlite3.html#using-the-connection-as-a-context-manager , though the last time I did this sort of crud with sqlite3 and without sqlalchemy I wrote my own `_execute` and `_query` wrappers

Reply:
Is there a real reason to do this other than to not have to type `db.commit()`? It's not particularly shorter, or anything. Is it just more idiomatic?

Reply:
it automatically rolls back on failure, though I guess since you create a new cursor every time there isn't much benefit